In [1]:
# Import necessary libraries
import pandas as pd

# Load the cleaned dataset from the processed folder
processed_data_path = "C:/project/Freight_Analysis_Framework-/data/processed/freight_analysis_cleaned.csv"
regional_data_cleaned = pd.read_csv(processed_data_path)

# Check the first few rows of the loaded data
regional_data_cleaned.head()


fr_orig  dms_orig  dms_dest  fr_dest  fr_inmode  dms_mode  fr_outmode  \
0    804.0        11        11    801.0        3.0         1         4.0   
1    804.0        11        19    801.0        3.0         1         4.0   
2    804.0        11       129    801.0        3.0         1         4.0   
3    804.0        11       131    801.0        3.0         1         4.0   
4    804.0        11       139    801.0        3.0         1         4.0   

   sctg2  trade_type  tons_2012  ...  tmiles_2015  tmiles_2020  tmiles_2025  \
0      1           1     0.3231  ...       0.3817       0.4318       0.4791   
1      1           1     0.3231  ...      25.7995      27.4014      28.5980   
2      1           1     0.3231  ...       0.5211       0.5033       0.4917   
3      1           1     0.3231  ...       1.4469       1.3902       1.4049   
4      1           1     0.3231  ...       0.7550       0.6872       0.6569   

   tmiles_2030  tmiles_2035  tmiles_2040  tmiles_2045  curval_2013  \
0       0.5265       0.5854       0.6449       0.6656      36.7205   
1      29.6591      30.9511      32.5587      33.8550     283.6623   
2       0.4819       0.4688       0.4860       0.5272       1.3439   
3       1.4619       1.5058       1.5015       1.6357      10.7475   
4       0.6488       0.6353       0.6072       0.6476       4.5875   

   curval_2014  curval_2015  
0      46.3411      42.7443  
1     333.8676     301.8007  
2       1.4783       1.3083  
3      10.9878       9.6642  
4       4.5780       3.9517  

[5 rows x 42 columns]

Observations:
Columns:

The dataset has 42 columns, including freight volume, freight value, and various other columns like mode of transport, origin/destination, and yearly tons.
Data Types:

The data contains a mix of numerical columns (float64, int64), and categorical columns (fr_orig, fr_dest, fr_inmode, etc.), which will need to be processed for feature engineering.
Missing Data:

It appears that we've already handled missing data in the cleaning process, so now we can move forward with generating features from the existing data.


In [2]:
# Create yearly growth features for tons and values
regional_data_cleaned['Volume_Growth_2012_2013'] = regional_data_cleaned['tons_2013'] / regional_data_cleaned['tons_2012'] - 1
regional_data_cleaned['Volume_Growth_2013_2014'] = regional_data_cleaned['tons_2014'] / regional_data_cleaned['tons_2013'] - 1
regional_data_cleaned['Volume_Growth_2014_2015'] = regional_data_cleaned['tons_2015'] / regional_data_cleaned['tons_2014'] - 1

regional_data_cleaned['Value_Growth_2012_2013'] = regional_data_cleaned['value_2013'] / regional_data_cleaned['value_2012'] - 1
regional_data_cleaned['Value_Growth_2013_2014'] = regional_data_cleaned['value_2014'] / regional_data_cleaned['value_2013'] - 1
regional_data_cleaned['Value_Growth_2014_2015'] = regional_data_cleaned['value_2015'] / regional_data_cleaned['value_2014'] - 1

# Display the new columns
regional_data_cleaned[['Volume_Growth_2012_2013', 'Volume_Growth_2013_2014', 'Volume_Growth_2014_2015',
                       'Value_Growth_2012_2013', 'Value_Growth_2013_2014', 'Value_Growth_2014_2015']].head()


Volume_Growth_2012_2013  Volume_Growth_2013_2014  Volume_Growth_2014_2015  \
0                55.597029                 0.082657                 0.035943   
1               674.192820                 0.009734                 0.015244   
2                 1.745280                -0.056257                -0.006092   
3                19.119777                -0.122941                -0.012172   
4                 7.887960                -0.143852                -0.030546   

   Value_Growth_2012_2013  Value_Growth_2013_2014  Value_Growth_2014_2015  
0               15.130461                0.082660                0.035944  
1              123.606288                0.009734                0.015244  
2               -0.182785               -0.056331               -0.006020  
3                3.721146               -0.122925               -0.012182  
4                1.015184               -0.143877               -0.030534

Volume Growth:

The growth in freight volume varies significantly across rows. For instance:
In 2012-2013, Volume_Growth_2012_2013 shows values as high as 674% for certain records. This is likely a case of large shipments in those years.
Some entries show negative growth (e.g., -0.056257 in Volume_Growth_2013_2014), indicating a decrease in freight volume in some cases.
Value Growth:

The Value Growth columns exhibit similar variability. The values in the 2012-2013 period show large jumps, such as 123.61% for certain rows, suggesting that value could be more volatile due to factors like price fluctuations or commodity changes.
Growth Interpretation:

The positive growth values suggest that freight volume and value increased in those years, which is typical for expanding markets or industries.
Negative growth values indicate periods of decline, which could be due to changes in demand, market conditions, or industry-specific factors.
These growth features are essential in understanding the dynamics and trends in freight movements over time.


In [3]:
import numpy as np

# Apply log transformation to the freight volume and value columns
regional_data_cleaned['Log_Freight_Volume_2012'] = np.log1p(regional_data_cleaned['tons_2012'])
regional_data_cleaned['Log_Freight_Value_2012'] = np.log1p(regional_data_cleaned['value_2012'])

# Display the transformed columns
regional_data_cleaned[['Log_Freight_Volume_2012', 'Log_Freight_Value_2012']].head()


Log_Freight_Volume_2012  Log_Freight_Value_2012
0                 0.279977                1.146450
1                 0.279977                1.146450
2                 0.279977                0.936485
3                 0.279977                1.146450
4                 0.279977                1.146450

Log Transformation:

The log transformation applied to both freight volume (tons_2012) and freight value (value_2012) has resulted in values that are now on a logarithmic scale. This transformation is particularly useful for right-skewed data, which seems to be the case with your original columns.
Log Freight Volume:

The transformed values for Log_Freight_Volume_2012 show much smaller values than the original data. This is expected, as logarithmic scaling reduces the impact of very large values, making the data more manageable for models that assume normality.
Log Freight Value:

Similar behavior is observed for Log_Freight_Value_2012, where values are transformed to a smaller scale.
Interpretation:

The log-transformed data is now more likely to follow a normal distribution or at least be more evenly distributed compared to the original skewed data. This will help in improving the performance of machine learning models that are sensitive to highly skewed data.


In [4]:
# Create interaction features between freight volume and value for 2012
regional_data_cleaned['Volume_Value_Interaction_2012'] = regional_data_cleaned['tons_2012'] * regional_data_cleaned['value_2012']
regional_data_cleaned['Log_Volume_Value_Interaction_2012'] = regional_data_cleaned['Log_Freight_Volume_2012'] * regional_data_cleaned['Log_Freight_Value_2012']

# Display the new interaction features
regional_data_cleaned[['Volume_Value_Interaction_2012', 'Log_Volume_Value_Interaction_2012']].head()


Volume_Value_Interaction_2012  Log_Volume_Value_Interaction_2012
0                       0.693696                           0.320980
1                       0.693696                           0.320980
2                       0.501128                           0.262195
3                       0.693696                           0.320980
4                       0.693696                           0.320980

Volume-Value Interaction Features:

The interaction feature Volume_Value_Interaction_2012 is the product of freight volume and freight value. This feature captures the relationship between the two variables and how they jointly contribute to freight movements in 2012.
For example, a value like 0.693696 means that the combination of volume and value for this specific row results in this interaction value.
Log-Transformed Interaction:

The log-transformed interaction feature Log_Volume_Value_Interaction_2012 represents the product of the log-transformed volume and value. This transformation helps deal with the skewness in the original data and makes the relationship between volume and value more interpretable.
The values here are also on a smaller scale, which is expected from log transformations.
Importance of Interaction Features:

These interaction features are important because they allow models to capture non-linear relationships between freight volume and value, which could be more predictive in tasks like regression or forecasting.


In [5]:
# Binning Freight Volume (tons_2012) into categories: Low, Medium, High
volume_bins = [0, 5000, 20000, 80000]  # Define the bin edges
volume_labels = ['Low', 'Medium', 'High']  # Bin labels
regional_data_cleaned['Volume_Category_2012'] = pd.cut(regional_data_cleaned['tons_2012'], bins=volume_bins, labels=volume_labels)

# Binning Freight Value (value_2012) into categories: Low, Medium, High
value_bins = [0, 5000, 20000, 80000]  # Define the bin edges
value_labels = ['Low', 'Medium', 'High']  # Bin labels
regional_data_cleaned['Value_Category_2012'] = pd.cut(regional_data_cleaned['value_2012'], bins=value_bins, labels=value_labels)

# Display the new categorical features
regional_data_cleaned[['Volume_Category_2012', 'Value_Category_2012']].head()


Volume_Category_2012 Value_Category_2012
0                  Low                 Low
1                  Low                 Low
2                  Low                 Low
3                  Low                 Low
4                  Low                 Low

In [6]:
# One-Hot Encoding for the mode of transport columns (fr_inmode and fr_outmode)
regional_data_cleaned = pd.get_dummies(regional_data_cleaned, columns=['fr_inmode', 'fr_outmode'], drop_first=True)

# Display the new one-hot encoded columns
regional_data_cleaned.head()


fr_orig  dms_orig  dms_dest  fr_dest  dms_mode  sctg2  trade_type  \
0    804.0        11        11    801.0         1      1           1   
1    804.0        11        19    801.0         1      1           1   
2    804.0        11       129    801.0         1      1           1   
3    804.0        11       131    801.0         1      1           1   
4    804.0        11       139    801.0         1      1           1   

   tons_2012  tons_2013  tons_2014  ...  fr_inmode_4.0  fr_inmode_5.0  \
0     0.3231    18.2865    19.7980  ...          False          False   
1     0.3231   218.1548   220.2783  ...          False          False   
2     0.3231     0.8870     0.8371  ...          False          False   
3     0.3231     6.5007     5.7015  ...          False          False   
4     0.3231     2.8717     2.4586  ...          False          False   

   fr_inmode_6.0  fr_inmode_7.0  fr_outmode_2.0  fr_outmode_3.0  \
0          False          False           False           False   
1          False          False           False           False   
2          False          False           False           False   
3          False          False           False           False   
4          False          False           False           False   

   fr_outmode_4.0  fr_outmode_5.0  fr_outmode_6.0  fr_outmode_7.0  
0            True           False           False           False  
1            True           False           False           False  
2            True           False           False           False  
3            True           False           False           False  
4            True           False           False           False  

[5 rows x 64 columns]

Binned Categories:

Volume_Category_2012:
All the first few rows fall into the Low category, indicating that for many of these records, the freight volume in 2012 is low (less than 5,000 tons).
Value_Category_2012:
Similarly, the freight value for these rows is also Low, suggesting that these records have freight values below 5,000.
One-Hot Encoding:

The columns fr_inmode and fr_outmode have been one-hot encoded into binary features for each category.
For example, fr_inmode_4.0 indicates whether the record has a mode of transport 4.0.
The False/True values indicate whether a particular mode of transport is used for the given record.
Resulting Columns:

After one-hot encoding, the number of columns has increased to 64. This includes the original columns plus the new categorical features from binning and the one-hot encoded features.


In [7]:
# Check the final dataset overview
print("Final Dataset Info:")
print(regional_data_cleaned.info())

# Display a few rows of the final dataset to ensure all transformations have been applied
print("\nFinal Dataset Head:")
print(regional_data_cleaned.head())


Final Dataset Info:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1660972 entries, 0 to 1660971
Data columns (total 64 columns):
 #   Column                             Non-Null Count    Dtype   
---  ------                             --------------    -----   
 0   fr_orig                            1660972 non-null  float64 
 1   dms_orig                           1660972 non-null  int64   
 2   dms_dest                           1660972 non-null  int64   
 3   fr_dest                            1660972 non-null  float64 
 4   dms_mode                           1660972 non-null  int64   
 5   sctg2                              1660972 non-null  int64   
 6   trade_type                         1660972 non-null  int64   
 7   tons_2012                          1660972 non-null  float64 
 8   tons_2013                          1660972 non-null  float64 
 9   tons_2014                          1660972 non-null  float64 
 10  tons_2015                          1660972 non-null  float

Non-Null Counts:

All columns are now non-null, indicating that we've successfully handled missing values earlier in the process. This is a good sign that our data is clean and ready for modeling.
Number of Columns:

The final dataset has 64 columns, including the original columns and the new features created during the feature engineering process (growth features, log-transformed features, interaction features, and the one-hot encoded columns).
Data Types:

float64 for most columns (indicating continuous numerical data).
int64 for categorical variables (e.g., dms_orig, dms_dest, trade_type).
The one-hot encoded columns are of boolean type (True/False).
Final Dataset Ready for Modeling:

The dataset is now ready for machine learning tasks, such as regression, classification, or clustering, depending on the problem you're working on.


In [8]:
import os

In [9]:
# Save the feature-engineered dataset to the processed folder
feature_engineered_data_path = "C:/project/Freight_Analysis_Framework-/data/processed/freight_analysis_feature_engineered.csv"
regional_data_cleaned.to_csv(feature_engineered_data_path, index=False)
print(f"Feature-engineered data saved successfully to {feature_engineered_data_path}")


Feature-engineered data saved successfully to C:/project/Freight_Analysis_Framework-/data/processed/freight_analysis_feature_engineered.csv
